### DISCUSSION TITLE
`git log | bat` looks nice, but `PAGER=bat git log` looks bland

### DISCUSSION THREAD

**Discussion by davidmaxwaterman** [sharkdp/bat#2329] [Q&A]
(Sorry, I'm new to bat and OSX)

If I do `git log | bat`, I get nice colours, notably 'commit', 'Author', and 'Date' are in red, hash is purple, comment in orange. eg:

![screenshot](https://user-images.githubusercontent.com/975698/193013400-a67db353-8fbf-4d1d-bd2b-3db609440c30.png)

However, if I do `git -c core.pager=bat log`, it only has very limited colours...`commit` and the hash itself at yellow, the branch name (eg `HEAD ->`) are cyan, and any local branch name is green; but more importantly, the rest of the commit is some kind of ugly grey colour:

![screenshot-1](https://user-images.githubusercontent.com/975698/193014391-84dfbab9-fc80-461c-8fbe-a5946ec8ee9a.png)

I just noticed that `$ git config -l --global` also uses bat, but the colour highlighting is all 'bland'

Any ideas what's going on?

```
 $ bat --version                                                                                            
bat 0.22.1
$ brew info bat                                                                                            
==> bat: stable 0.22.1 (bottled), HEAD
Clone of cat(1) with syntax highlighting and Git integration
https://github.com/sharkdp/bat
/opt/homebrew/Cellar/bat/0.22.1 (14 files, 4.7MB) *
  Poured from bottle on 2022-09-29 at 11:20:43
From: https://github.com/Homebrew/homebrew-core/blob/HEAD/Formula/bat.rb
License: Apache-2.0 or MIT
==> Dependencies
Build: rust ✘
==> Options
--HEAD
	Install HEAD version
==> Caveats
zsh completions have been installed to:
  /opt/homebrew/share/zsh/site-functions
==> Analytics
install: 44,926 (30 days), 68,454 (90 days), 220,976 (365 days)
install-on-request: 44,576 (30 days), 67,829 (90 days), 219,032 (365 days)
build-error: 0 (30 days)
$ git --version                                                                                            
git version 2.35.1
$ printenv | grep PAGER                                                                                    
PAGER=bat
```

**Comment by keith-hall** [comment]
You can try telling `bat` the language to highlight with explicitly, i.e. `git -c core.pager="bat -l'Git Log'" log`

**Reply by davidmaxwaterman** [reply]
Thanks for responding.

That is interesting...it does change things, but it is a sort of 'in between'.
It seems to change the colours of the first commit's 'header lines', and even the first line of the comment 'Bug .... - ...', plus it seems to highlight bug numbers and things that I presume it thinks are bug numbers.

Below are a few larger screenshots of the same log.

I notice that the actual text is different. When using a pipe, the log is simpler (doesn't output the branches/tags). I've seen that before - due to the git log `--decorate=auto` default changing when the output is a terminal vs a pipe. Specifying a decorate of `short` or `full` changes things yet again. Could it simply be poor syntax pattern matching? If I specify decorate of 'no' for both...(trying it out)...well, I was hoping that would at least make them the same, but it doesn't - the text is the same though, so I think we can rule out that being the issue :/

FWIW, the command I'll test with from now on will specify --decorate, eg:

```
git log --decorate=no | bat
git -c core.pager=bat log --decorate=no
```

(to be clear, the screenshots don't specify, and specifying the language still does the 'in between' highlighting show below)

Note that the highlighting of the word 'error' is done by iTerm2, not bat - and turning that off makes no difference to the problem.

With `git -c core.pager="bat -l'Git Log'" log`:

![bat-l-git-log](https://user-images.githubusercontent.com/975698/193231167-36da330a-9ab4-4fe5-ab97-17cb76ecc34d.png)

and with `git log | bat`:

![bat-pipe](https://user-images.githubusercontent.com/975698/193231227-bc0793e2-ad3e-4ef8-a188-bd380b15499d.png)

and here is the same with `git -c core.pager="bat" log`:

![Screenshot 2022-09-30 at 09 47 29](https://user-images.githubusercontent.com/975698/193231514-59163af0-b41d-441a-a454-b372ece63f75.png)

**Reply by keith-hall** [reply]
I think that this bug is related: https://github.com/sharkdp/bat/issues/1632
I have made a fix upstream, but I think `bat` is stuck on an old version until the syntax highlighting library we use is updated to support more recent features: https://github.com/trishume/syntect/issues/323

**Reply by davidmaxwaterman** [reply]
OK. Yes, it does look related, though not precisely the same.

TBH, I'm struggling to make bat 'look good' in many scenarios. Even `man git-log` was difficult to read (`PAGER=bat`) - or perhaps that's more a matter of the default theme and I should just pick another one? Using TwoDark is readable, but bland/monochrome (with some seemingly random highlighting). Any recommendation?

**Comment by davidmaxwaterman** [comment]
You know, thinking about it, I don't see how it can be due to that bug you mention. If I do `git log | bat`, it looks fine...with the same input...so how can it be?

Somehow, when run through a config (eg with `PAGER=bat` or `core.pager=bat`) it doesn't look good; but with it as a filter ie `| bat` it *does* look good.

So, it seems like it is doing something different between those two cases.

**Comment by davidmaxwaterman** [comment]
OK, I found something.

If I use the following git command then it works the same with either `PAGER=bat` or `| bat`:

```
git log --no-color --no-decorate
```

However, `--no-decorate` means the output isn't as useful (due to no branch/tags/etc).

If I use `PAGER="tee ~/tmp.txt"` and look at the raw output, I see:

The `--no-decorate` produces a `commit` line like this:

```
^[[33mcommit f4f084dd1ed256101665becb21828ef8d594e399^[[m
```

and the `--no-color` produces:

```
commit f4f084dd1ed256101665becb21828ef8d594e399 (HEAD -> esr102, origin/branches/default/5909e8822d62ab7f80d6667e448742396715f207, origin/bookmarks/esr102)
```

So, the escape sequences in the --no-decorate output is messing bat up, and
the decoration text in the --no-color output is messing bat up.

The only way to get good output for both PAGER and pipe, is to use both --no-decorate and --no-color.

So, it is some additional differences in the output of `git log` between sending output to terminal and sending into a pipe, that are causing this issue?

**Reply by cben** [reply]
IIUC git's default is `--color=auto` so redirecting stdout e.g. `git log | bat` implies `--no-color` as git can't know who'll be reading.  
OTOH a pager (whether set by core.pager config, or GIT_PAGER env var) is expected to support color (like `less -R`).

I see how having both git & bat colorize will collide, unpredictably.
But for me `GIT_PAGER="bat -l'gitlog'" git log --no-color --decorate` works fine, as of bat 0.25.0.  

Though it's not as good as git's own colorization, e.g. git also colors decorations, and --graph lines, and honors `%C` color specs in custom --pretty formats...  Git really knows what it's printing while bat has to reverse-engineer it.

Are you sure you _want_ bat to parse & colorize git output?  
One area where bat _could_ win is syntax highlighting any _source code_ that's printed but it's tricky to configure, and the central use case of `git diff` (and `git log --patch`) is covered by https://github.com/dandavison/delta/ (which uses same highlighting as bat +-backgrounds :rainbow:)

**Comment by drskoolie** [comment]
Having bat as my pager messes up my other aliases, so I put this in my .gitconfig:

`
[alias]

	lo = "! f() { git log --no-color --no-decorate | bat -n -l 'Git log' ; }; f"
`

And it is run by typing in `git lo`
